plugins/generator-1.18.2/forge-1.18.2/templates/enchantment.java.ftl

```diff
@@ -59,16 +59,40 @@ public class ${name}Enchantment extends Enchantment {
 
 	<#if data.compatibleEnchantments?has_content>
 		@Override protected boolean checkCompatibility(Enchantment ench) {
-		    return <#if data.excludeEnchantments>!</#if>List.of(
-                <#list data.compatibleEnchantments as compatibleEnchantment>${compatibleEnchantment}<#sep>,</#list>).contains(ench);
+			return <#if data.excludeEnchantments>!</#if>List.of(
+				<#list data.compatibleEnchantments as compatibleEnchantment>${compatibleEnchantment}<#sep>,</#list>).contains(ench);
 		}
 	</#if>
 
 	<#if data.compatibleItems?has_content>
 		@Override public boolean canApplyAtEnchantingTable(ItemStack stack) {
 			Item item = stack.getItem();
-			return <#if data.excludeItems>!</#if>List.of(
-                <#list data.compatibleItems as compatibleItem>${mappedMCItemToItem(compatibleItem)}<#sep>,</#list>).contains(item);
+			<#assign items = []>
+			<#assign tags = []>
+			<#list data.compatibleItems as compatibleItem>
+				<#if compatibleItem.getUnmappedValue().startsWith("TAG:")>
+					<#assign tags += [compatibleItem.getUnmappedValue().replace("TAG:", "")]>
+				<#elseif generator.map(compatibleItem.getUnmappedValue(), "blocksitems", 1).startsWith("#")>
+					<#assign tags += [generator.map(compatibleItem.getUnmappedValue(), "blocksitems", 1).replace("#", "")]>
+				<#else>
+					<#assign items += [mappedMCItemToItem(compatibleItem)]>
+				</#if>
+			</#list>
+
+			return <#if data.excludeItems>!</#if>
+				List.of(

```

Comment:
This will still keep empty List.of() if only tags are defined. Did you consider using the same approach as in eg. armor with `Ingredient` class and then you can use `Ingredient#test` to test if stack matches Ingredient